src/app/models/__tests__/form.server.model.spec.ts

```diff
@@ -1545,5 +1545,61 @@ describe('Form Model', () => {
         expect(actual).toBeInstanceOf(mongoose.Error.ValidationError)
       })
     })
+
+    describe('reorderFormFieldById', () => {
+      let form: IFormSchema
+      const FIELD_ID_TO_REORDER = new ObjectId().toHexString()
+
+      beforeEach(async () => {
+        form = await Form.create({
+          admin: populatedAdmin._id,
+          responseMode: ResponseMode.Email,
+          title: 'mock email form',
+          emails: [populatedAdmin.email],
+          form_fields: [
+            generateDefaultField(BasicField.Checkbox),
+            generateDefaultField(BasicField.HomeNo, {
+              title: 'some mock title',
+              _id: FIELD_ID_TO_REORDER,
+            }),
+            generateDefaultField(BasicField.Email),
+          ],
+        })
+      })
+
+      it('should return updated form with reordered fields successfully', async () => {
+        // Act
+        const originalFields =
+          cloneDeep(
+            (form.form_fields as Types.DocumentArray<IFieldSchema>).toObject(),
+          ) ?? []
+        const updatedForm = await form.reorderFormFieldById(
+          FIELD_ID_TO_REORDER,
+          0,
+        )
+
+        // Assert
+        expect(updatedForm).not.toBeNull()
+        expect(
+          (updatedForm?.form_fields as Types.DocumentArray<IFieldSchema>).toObject(),
+        ).toEqual([
+          // Should be rearranged to the 0th index position, and the previously
+          // 0th index field should be pushed to 1st index.
+          originalFields[1],
+          originalFields[0],
+          originalFields[2],
+        ])
+      })
+
+      it('should return null if given fieldId is invalid', async () => {
```

Comment:
do we need a test for when the new position is greater than the length of the array?

Reply:
nice catch, i thought there were already tests for `immutable-array-fns`. Will add them. 

for now, the newPosition is determined by slice and can be negative or positive etc, slice will either slice backwards or slice to the end. do you think this should be the behaviour or do we need guards for that?

Reply:
I remember we had this discussion the first time you implemented `immutable-array-fns`, and I think we concluded that the functions should just be dumb and it should be the responsibility fo the caller to validate the `newIndex`. in this case we already have Joi validation to ensure that the new index is >0, which I think is sufficient to cover the negative case, which is why I figured we should just add tests for the >length case.

idt we need additional tests for the negative case. it's like any other Joi validation, eg we don't have tests for when a service is called with an invalid fieldType because we assume the fieldType has been Joi-validated

Reply:
noted, will add tests for the newPosition > fields.length. (but yea, since immutable-array-fns is using slice under the hood, will be safe)

Reply:
Resolved in 7f46d72